File: Scala/src/Sources/Quick/quick.scala
package Sources.Quick

import Others.swap.swap
import Others.value._

object quick {
    private def partition(a: Array[Int], left: Int, right: Int): Int = {
        var l = left
        var r = right
        val p = a(l + (r - l) / 2)

        while (l <= r) {
            while (a(l) < p)
                l += 1
            while (a(r) > p)
                r -= 1
            if (l <= r) {
                swap(a, l, r)
                l += 1
                r -= 1
            }
        }
        l
    }

    private def quick_sort(a: Array[Int], l: Int, r: Int): Unit = {
        val pl = partition(a, l, r)

        if (l < pl - 1)
            quick_sort(a, l, pl - 1)
        if (r > pl)
            quick_sort(a, pl, r)
    }

    def main(args: Array[String]): Unit = {
        val s = 1000000 // Changeable size
        val a = rev_p(s)

        val start = System.currentTimeMillis()

        quick_sort(a, 0, a.length - 1)

        val end = System.currentTimeMillis() - start

        for (i <- a.indices)
            println(a(i))
        print(end + " ms")
    }
}

File: Kotlin/src/main/kotlin/Sources/Quick/quick.kt
package Sources.Quick

import swap

fun partition(a: IntArray, l: Int, r: Int): Int {
    var l_ = l
    var r_ = r
    val p = a[l+(r-l)/2]

    while (l_ <= r_) {
        while (a[l_] < p)
            ++l_
        while (a[r_] > p)
            --r_
        if (l_ <= r_) {
            swap(a, l_, r_)
            ++l_
            --r_
        }
    }
    return l_
}

fun quick_sort(a: IntArray, l: Int, r: Int) {
    val pl = partition(a, l, r)

    if (l < pl-1)
        quick_sort(a, l, pl-1)
    if (r > pl)
        quick_sort(a, pl, r)
}

fun main() {
    val a = intArrayOf(5, 4, 3, 2, 1)

    val start = System.currentTimeMillis()

    quick_sort(a, 0, a.size-1)

    val end = System.currentTimeMillis() - start

    for (i in a.indices)
        println(a[i])
    print("$end ms")
}